app/scripts/lib/transaction/smart-transactions.ts

```diff
@@ -0,0 +1,339 @@
+import SmartTransactionsController from '@metamask/smart-transactions-controller';
+import {
+  Fee,
+  Fees,
+  SmartTransactionStatuses,
+  SmartTransaction,
+} from '@metamask/smart-transactions-controller/dist/types';
+import type { Hex } from '@metamask/utils';
+import {
+  TransactionController,
+  TransactionMeta,
+  TransactionParams,
+} from '@metamask/transaction-controller';
+import log from 'loglevel';
+import {
+  RestrictedControllerMessenger,
+  EventConstraint,
+} from '@metamask/base-controller';
+import {
+  AddApprovalRequest,
+  UpdateRequestState,
+  StartFlow,
+  EndFlow,
+} from '@metamask/approval-controller';
+
+import { decimalToHex } from '../../../../shared/modules/conversion.utils';
+import { CANCEL_GAS_LIMIT_DEC } from '../../../../shared/constants/smartTransactions';
+import {
+  SMART_TRANSACTION_CONFIRMATION_TYPES,
+  ORIGIN_METAMASK,
+} from '../../../../shared/constants/app';
+
+const namespace = 'SmartTransactions';
+
+type AllowedActions =
+  | AddApprovalRequest
+  | UpdateRequestState
+  | StartFlow
+  | EndFlow;
+
+export type SmartTransactionsControllerMessenger =
+  RestrictedControllerMessenger<
+    typeof namespace,
+    AllowedActions,
+    EventConstraint,
+    AllowedActions['type'],
+    never
+  >;
+
+export type SubmitSmartTransactionRequest = {
+  transactionMeta: TransactionMeta;
+  smartTransactionsController: SmartTransactionsController;
+  transactionController: TransactionController;
+  isSmartTransaction: boolean;
+  controllerMessenger: SmartTransactionsControllerMessenger;
+  featureFlags: Record<string, any>;
+};
+
+export class SmartTransactionHook {
```

Comment:
I like the idea that if a class has only 1 public function, we can just export a simplified wrapper like this:
```
export const submitSmartTransactionHook = (
  request: SubmitSmartTransactionRequest,
) => {
  const smartTransactionHook = new SmartTransactionHook();
  return smartTransactionHook.submit(request);
};
```

However, there is one more public function in the class called `onApproveOrReject`, which is used in tests to simulate approval or reject calls from the `ApprovalController:addRequest` function. If there is some other way to simulate it in tests at the right moment without adding any artificial delays, so I could only have one public class function, then I can use the simplified wrapper.

Reply:
Ideally, we want to avoid making methods public only for the benefit of unit tests.

In this scenario, we already have the mock `controllerMessenger` instance in the tests, so can just mock it to resolve with a suitable promise when `call` is invoked with `ApprovalController:addRequest`.